text/0000-router-service-computed-query-params.md

```diff
@@ -0,0 +1,324 @@
+
+Everyone has different query param serialization and deserialization needs depending on a variety of factors.
+
+By default, the query params will be serialized and deserialized via the builtin URLSearchParams API, and polyfilled for IE11.
+
+Should someone decide to customize how serilaization and deserializion transforms the query params, that can be done directly on the router:
+
+```ts
+import EmberRouter from '@ember/routing/router';
+
+import config from '../config/environment';
+
+const Router = EmberRouter.extend({
+  location: config.locationType,
+  rootURL: config.rootURL,
+
+  queryParamsConfig: {
+    serialize(queryParams: object): string {
+      // serialize object for query string
+      // default to URLSearchParams, polyfilled for IE11
+    },
+    deserialize(queryString: string): object {
+      // parse to object for use in `injectedRouter.queryString`
+      // also default to URLSerachParams
+    }
+  }
+});
+```
+
+
+This will address a long standing issues from as far back as 2016,
+some new functionality for serialization and deserialization could be powered by [qs](https://www.npmjs.com/package/qs) ([3.4kb (gzip+min)](https://bundlephobia.com/result?p=qs@6.7.0)) or a lternatively, [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) -- this would enable the setting of arrays and objects which is not possible today.
+
+
+### Sticky Query Params
+
+By default, transitionTo will clear the query params, unless specified inside the transiion.
+
+If query params are defined ahead of time as sticky, they will persist in the URL between sub routes.
+
+This can be configured in the `Router.map` function:
+```ts
+Router.map(function() {
+  this.queryParams('bar');
+
+  this.route('faq');
+
+  this.route('posts', function() {
+    this.queryParams('foo', 'baz');
+
+    this.route('new');
+    this.route('index');
+    this.route(
+      'show',
+      { path: '/:postId', queryParams: ['hideComments', 'invertColors'] },
+      function() {
+        this.route('edit');
+        this.route('comment');
+        this.route('share');
+      }
+    );
+  });
+});
+```
+
+This method of dealing with query params implies the following:
+ - All query params, even if not specified are allowed. given the above example, I could use the qp "strongestAvenger" anywhere
+ - Any sort of transition will clear the query params, unless it is defined as a sticky queryParam. So, if I'm on the posts/new route with the query params "foo" and "baz", and transition to posts/show, those query params are still available. If I navigate to the faq page, foo and baz are cleared from the URL.
+ - The globally defined query params will stick around until cleared manually. If I visit faq?bar=2, and then transition to posts. the bar=2 query param will still be present.
+ - The `this.queryParams` function will be available at every nesting of the route tree, but `queryParams` are also configurable in the route options hash.
```

Comment:
This is a little concerning. I'd prefer there to be only one way to do this. Maybe it should always be as an option on `this.route()`, with an option for whether or not it cascades down to child routes?

```js
this.route('search',
  {
    queryParams: {
      names: ['foo', 'bar', baz'],
      cascades: false
    }
}
```

Reply:
I'd also prefer to avoid adding `this.queryParams` to the context of each callback on the `Router.map`. Passing configured query params as metadata seems better (and allows better interop with other things that _also_ want to be router map metadata).

Reply:
Another open question from `this.queryParams()` is the order of declaration and if that matters. Since the order of declaration of `this.route()` *does* matter, I'd want a spec for what these variations do:

```js
this.queryParams('search');
this.route('foo');
```

vs:

```js
this.route('foo');
this.queryParams('search');
```

Reply:
@rwjblue @mehulkar
> I'd prefer there to be only one way to do this.

without `this.queryParams`, there'd be no way to set global query params, unless we started adding `this.route('application')` by default -- which may be fine

>  order of declaration and if that matters. 

there is no difference in order with `this.queryParams` as it's just a function-scoped invocation, and applies to the route-scope of `this`  (application at the top level)